### PR TITLE
change event channel to non-blocking channel

### PR DESCRIPTION
**Pull request by kadisi**
Signed-off-by: zhangjie <iamkadisi@163.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind test
> /kind failing-test

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
change event channel to non-blocking channel to Increase processing speed

### CODE REVIEW

**Comment by rohitsardesai83** [comment]
@kadisi , thanks for the contribution , can you please check the CI failures ?

**Comment by kadisi** [comment]
does this ci have `/retest` function like kubernetes?

**Comment by sids-b** [comment]
can we have the same change in all the managers in controller and deviceController ??

**Comment by kadisi** [comment]
@sids-b , ok, i think we need change. but is it ok to set channel buff to 20?

**Comment by sids-b** [comment]
I am not sure, Maybe make it configuration based and set default to 10 ?

**Comment by sids-b** [comment]
It maybe not be a required config, we have many such fields in controller. You can have a look at the config folder of controller.

**Comment by kadisi** [comment]
@sids-b think you,

**Comment by kadisi** [comment]
@sids-b by the way, what is the deference between `-` and `.` in the Key  when i use `config.CONFIG.GetValue()`.  i know `.` come from yaml config. and  does `-` come from cmd or ENV ?

**Comment by m1093782566** [comment]
Sorry, I am not against chaging the size of channle buffer but I need to know the effect after changing the size. Does it help improving the performance? We need to see the data.

**Comment by kadisi** [comment]
@m1093782566 do you have some comparative suggestions about test case.

**Comment by m1093782566** [comment]
I think we care about the e2e performance data comparisions. @edisonxiang is working on e2e performance test, you may ask him for suggestions.

**Comment by kadisi** [comment]
@m1093782566  thank you, @edisonxiang  I contact you by email

**Comment by edisonxiang** [comment]
Hey @kadisi Thanks your pr.
I guess this is something important for the performance.
We need to make more tests for the kubeedge performance.

Currently we only add channel buffer in the upstream like the following.
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/controller/controller/upstream.go#L93-L122
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/devicecontroller/controller/upstream.go#L64-L79

There are some places like below which is using  `make(chan watch.Event)`
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/controller/manager/configmap.go#L24
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/controller/manager/pod.go#L83-L84
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/controller/manager/secret.go#L24
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/devicecontroller/manager/device.go#L31
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/devicecontroller/manager/devicemodel.go#L31

Could you help to add the buffer config like below so that we can avoid to use the fixed value like `20`, perhaps we can set the default value like `1`. if the users want to change these buffer, they can modify the conf of kubeedge to adjust them.

https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/devicecontroller/config/buffer.go
https://github.com/kubeedge/kubeedge/blob/master/cloud/pkg/controller/config/buffer.go

@kadisi what do you think about that?
@rohitsardesai83 @m1093782566 @sids-b

**Comment by sids-b** [comment]
@kadisi , in a configuration file which is yaml, there is heirarchy. So ``` . ``` is used to traverse down the heirarchy. for eg:
```
kube:
  master: 127.0.0
```
So to get the value of master, you need to use the key kube.master as master is in the heirarchy of kube
As per my knowledge ``` - ``` doesn't hold any significance. If the key has a - for eg: master-ip, they we use it as kube.master-ip.

Also, environmental variable has higher preference than config file. If same key is present as environmental variable and in config file, The one in config file will be overwritten during runtime.

**Comment by kadisi** [comment]
@sids-b thank for your help.      @edisonxiang ok, i will set it

File: cloud/pkg/controller/config/buffer.go [hunk 1]
```diff
@@ -46,4 +55,26 @@ func init() {
 		QuerySecretBuffer = qsb
 	}
 	log.LOGGER.Infof("query secret buffer: %d", QuerySecretBuffer)
+
+	if peb, err := config.CONFIG.GetValue("pod-event-buffer").ToInt(); err != nil {
```
Comment (edisonxiang) [inline cloud/pkg/controller/config/buffer.go]: perhaps we need to replace 'pod-event-buffer' into 'controller.pod-event-buffer'
The below config should be that too.

File: cloud/pkg/devicecontroller/config/buffer.go [hunk 2]
```diff
@@ -9,11 +9,32 @@ import (
...
+
 func init() {
 	if psb, err := config.CONFIG.GetValue("devicecontroller.update-device-status-buffer").ToInt(); err != nil {
 		UpdateDeviceStatusBuffer = constants.DefaultUpdateDeviceStatusBuffer
 	} else {
 		UpdateDeviceStatusBuffer = psb
 	}
 	log.LOGGER.Infof("Update device status buffer: %d", UpdateDeviceStatusBuffer)
+
+	if deb, err := config.CONFIG.GetValue("device-event-buffer").ToInt(); err != nil {
```
Comment (edisonxiang) [inline cloud/pkg/devicecontroller/config/buffer.go]: perhaps we need to replace 'device-event-buffer' into 'devicecontroller.device-event-buffer'
The below config should be that too.

File: cloud/pkg/controller/config/buffer.go (hunk 1 above)
Comment (sids-b) [inline cloud/pkg/controller/config/buffer.go]: Can we refer to PR #490 and add these configurations under buffer for both controller and deviceController ?
Lets wait till pr #490 is approved though

Comment (kadisi) [reply cloud/pkg/controller/config/buffer.go]: ok

Comment (sids-b) [reply cloud/pkg/controller/config/buffer.go]: @kadisi , the PR is merged. Please take a look and add configurations under buffer

**Comment by rohitsardesai83** [comment]
@kadisi , can you please check ?

File: cloud/pkg/controller/config/buffer.go (hunk 1 above)
Comment (kadisi) [reply cloud/pkg/controller/config/buffer.go]: ok

File: cloud/pkg/devicecontroller/config/buffer.go (hunk 2 above)
Comment (kadisi) [reply cloud/pkg/devicecontroller/config/buffer.go]: ok

File: cloud/pkg/controller/manager/configmap.go [hunk 3]
```diff
@@ -1,6 +1,7 @@
 package manager
 
 import (
+	"github.com/kubeedge/kubeedge/cloud/pkg/controller/config"
```
[the remark this answers is not in the document]
Comment (kadisi) [reply cloud/pkg/controller/manager/configmap.go]: ok

File: cloud/pkg/controller/config/buffer.go (hunk 1 above)
Comment (kadisi) [reply cloud/pkg/controller/config/buffer.go]: ok

File: cloud/pkg/devicecontroller/config/buffer.go [hunk 4]
```diff
@@ -9,11 +9,32 @@ import (
...
 func init() {
 	if psb, err := config.CONFIG.GetValue("devicecontroller.buffer.update-device-status").ToInt(); err != nil {
 		UpdateDeviceStatusBuffer = constants.DefaultUpdateDeviceStatusBuffer
 	} else {
 		UpdateDeviceStatusBuffer = psb
 	}
-	log.LOGGER.Infof("Update device status buffer: %d", UpdateDeviceStatusBuffer)
+	log.LOGGER.Infof("devicecontroller Update device status buffer: %d", UpdateDeviceStatusBuffer)
+
+	if deb, err := config.CONFIG.GetValue("devicecontroller.device-event-buffer").ToInt(); err != nil {
```
Comment (edisonxiang) [inline cloud/pkg/devicecontroller/config/buffer.go]: As @sids-b suggests, it is better to add configurations under buffer like the line 19 mentions
"devicecontroller.buffer.update-device-status"
So the folloing modification is welcome:
Replace "devicecontroller.device-event-buffer" with "devicecontroller.buffer.device-event"
Replace "devicecontroller.device-model-event-buffer" with "devicecontroller.buffer.device-model-event"

File: cloud/pkg/controller/config/buffer.go [hunk 5]
```diff
@@ -18,32 +18,63 @@ var QueryConfigMapBuffer int
...
 
 	if qsb, err := config.CONFIG.GetValue("controller.buffer.query-secret").ToInt(); err != nil {
 		QuerySecretBuffer = constants.DefaultQuerySecretBuffer
 	} else {
 		QuerySecretBuffer = qsb
 	}
-	log.LOGGER.Infof("query secret buffer: %d", QuerySecretBuffer)
+	log.LOGGER.Infof("controller query secret buffer: %d", QuerySecretBuffer)
+
+	if peb, err := config.CONFIG.GetValue("controller.pod-event-buffer").ToInt(); err != nil {
```
Comment (edisonxiang) [inline cloud/pkg/controller/config/buffer.go]: The following modifications are welcome:
Replace "controller.pod-event-buffer" with "controller.buffer.pod-event"
Replace "controller.configmap-event-buffer" with "controller.buffer.configmap-event"
Replace "controller.secret-event-buffer" with "controller.buffer.secret-event"

@sids-b @kadisi How about your ideas?

Comment (sids-b) [reply cloud/pkg/controller/config/buffer.go]: Agree with @edisonxiang

File: cloud/pkg/controller/config/buffer.go [hunk 6]
```diff
@@ -18,32 +18,63 @@ var QueryConfigMapBuffer int
...
+var SecretEventBuffer int
+
 func init() {
 	if psb, err := config.CONFIG.GetValue("controller.buffer.update-pod-status").ToInt(); err != nil {
 		UpdatePodStatusBuffer = constants.DefaultUpdatePodStatusBuffer
 	} else {
 		UpdatePodStatusBuffer = psb
 	}
-	log.LOGGER.Infof("update pod status buffer: %d", UpdatePodStatusBuffer)
+	log.LOGGER.Infof("controller update pod status buffer: %d", UpdatePodStatusBuffer)
```
Comment (sids-b) [inline cloud/pkg/controller/config/buffer.go]: nit : Please start log with capital letters
applicable to similar places

File: cloud/pkg/controller/manager/secret.go [hunk 7]
```diff
@@ -1,6 +1,7 @@
 package manager
 
 import (
+	"github.com/kubeedge/kubeedge/cloud/pkg/controller/config"
```
Comment (sids-b) [inline cloud/pkg/controller/manager/secret.go]: nit: please seperate kubeedge and vendor packages with a blank line.

File: cloud/pkg/devicecontroller/config/buffer.go (hunk 4 above)
Comment (sids-b) [reply cloud/pkg/devicecontroller/config/buffer.go]: +1

File: cloud/pkg/controller/config/buffer.go (hunk 5 above)
Comment (kadisi) [reply cloud/pkg/controller/config/buffer.go]: ok

**Comment by kadisi** [comment]
@sids-b @edisonxiang

**Comment by m1093782566** [comment]
/lgtm

leave approve to @edisonxiang @sids-b

**Comment by edisonxiang** [comment]
@kadisi please rebase your pr, and then we will merge it. Thanks very much.

**Pull request merged**